uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/controllers/leave-request-popup-controller_test.js

```diff
@@ -554,6 +577,105 @@
+            it('should disable to date and to type', function () {
+              expect($ctrl.leaveRequest.from_date).toBeFalsy();
+            });
+          });
+
+          describe('after from date is selected', function () {
+            beforeEach(function () {
+              $ctrl.uiOptions.fromDate = new Date(date2017);
+              $ctrl.onDateChange($ctrl.uiOptions.fromDate, 'from');
+              $scope.$digest();
+            });
+
+            it('should enable to date and to type', function () {
+              expect($ctrl.leaveRequest.from_date).toBeTruthy();
+            });
+
+            it('should check if date is in any absence period', function () {
+              expect($ctrl.error).toBeNull();
+            });
+
+            it('should update calendar', function () {
+              expect(WorkPatternAPI.getCalendar).toHaveBeenCalled();
+            });
+
+            it('should not show balance', function () {
+              expect($ctrl.uiOptions.showBalance).toBeFalsy();
+            });
+
+            describe('from available absence period', function () {
+              var oldPeriodId;
+
+              beforeEach(function () {
+                $ctrl.uiOptions.toDate = null;
+                oldPeriodId = $ctrl.period.id;
+                $ctrl.uiOptions.fromDate = new Date(date2016);
+                $ctrl.onDateChange($ctrl.uiOptions.fromDate, 'from');
+                $scope.$digest();
+              });
+
+              it('should change absence period', function () {
+                expect($ctrl.period.id).not.toEqual(oldPeriodId);
+              });
+
+              it('should set min and max to date', function () {
+                expect($ctrl.uiOptions.date.to.options.minDate).not.toBeNull();
+                expect($ctrl.uiOptions.date.to.options.maxDate).not.toBeNull();
+              });
+
+              it('should update absence types from Entitlements', function () {
+                expect(EntitlementAPI.all).toHaveBeenCalled();
+              });
+
+              it('should not show balance', function () {
+                expect($ctrl.uiOptions.showBalance).toBeFalsy();
+              });
+
+              it('should reset to date', function () {
+                expect($ctrl.leaveRequest.to_date).toBeNull();
+              });
+            });
+
+            describe('from unavailable absence period', function () {
+              beforeEach(function () {
+                $ctrl.uiOptions.fromDate = new Date(date2013);
+                $ctrl.onDateChange($ctrl.uiOptions.fromDate, 'from');
+                $scope.$digest();
+              });
+
+              it('should show error', function () {
+                expect($ctrl.error).toEqual(jasmine.any(String));
+              });
+            });
+
+            describe('and to date is selected', function () {
+              beforeEach(function () {
+                setDates2016(true, true);
+                $scope.$digest();
```

Comment:
Do you need `$scope.$digest();`? We are calling that already inside `setDates2016`